src/components/Table/utils/TableState.ts

```diff
@@ -342,3 +401,59 @@ function flattenRows(rows: GridDataRow<any>[]): GridDataRow<any>[] {
   const childRows = rows.flatMap((r) => (r.children ? flattenRows(r.children) : []));
   return [...rows, ...childRows];
 }
+
+// Exported for testing purposes
+export function deriveSortState(
+  currentSortState: SortState | {},
+  clickedKey: string,
+  initialSortState: SortState | undefined,
+): SortState | undefined {
+  // If the current sort state is not defined then sort ASC on the clicked key.
+  if (!isSortState(currentSortState)) {
+    return { ...initialSortState, current: { columnId: clickedKey, direction: ASC } };
+  }
+
+  // const { sortedColumnId: currentKey, direction: currentDirection } = currentSortState;
+  const {
+    current: { columnId: currentKey, direction: currentDirection },
+  } = currentSortState;
+
+  // If clicking a new column, then sort ASC on the clicked key
+  if (clickedKey !== currentKey) {
+    return { ...initialSortState, current: { columnId: clickedKey, direction: ASC } };
+  }
+
+  // If there is an `initialSortState` and we're clicking on that same key, then flip the sort.
+  // Handles cases where the initial sort is DESC so that we can allow for DESC to ASC sorting.
+  if (initialSortState && initialSortState.current.columnId === clickedKey) {
+    return {
+      ...initialSortState,
+      current: { columnId: clickedKey, direction: (currentDirection as any as string) === ASC ? DESC : ASC },
+    };
+  }
+
+  // Otherwise when clicking the current column, toggle through sort states
+  if ((currentDirection as any as string) === ASC) {
+    // if ASC -> go to desc
+    return { ...initialSortState, current: { columnId: clickedKey, direction: DESC } };
+  }
+
+  // Else, direction is already DESC, so revert to original sort value.
+  return initialSortState;
+}
+
+type ColumnSort = {
+  columnId: string | undefined;
+  direction: Direction | undefined;
```

Comment:
Suggestion: Can these `| undefined`s go away, with the idea that maybe we don't have sort in place (so `tableState.sort = {} or undefined`) but if we do have current/persistent sort in place, then that sort should always have both an id and a direction.